tests/update_workflow_test.go

```diff
@@ -3918,3 +3919,471 @@ func (s *integrationSuite) TestUpdateWorkflow_CompletedSpeculativeWorkflowTask_D
+	_, err = poller.HandlePartialWorkflowTask(wt1Resp.GetWorkflowTask(), false)
+	s.NoError(err)
+	s.EqualValues(0, wt1Resp.ResetHistoryEventId)
+
+	// Get shardId so we can unload the shard later
+	ms, err := s.adminClient.DescribeMutableState(testCtx, &adminservice.DescribeMutableStateRequest{
+		Namespace: s.namespace,
+		Execution: tv.WorkflowExecution(),
+	})
+	s.NoError(err)
+	shardId, err := strconv.Atoi(ms.ShardId)
+	s.NoError(err)
+
+	// send update wf request, this will trigger speculative wft
+	go func() {
+		_, _ = s.sendUpdate(tv, "1")
+	}()
+
+	// poll the speculative wft
+	wft1, err1 := s.engine.PollWorkflowTaskQueue(testCtx, &workflowservice.PollWorkflowTaskQueueRequest{
+		Namespace: s.namespace,
+		TaskQueue: tv.TaskQueue(),
+	})
+	s.NoError(err1)
+	s.NotNil(wft1)
+	s.True(len(wft1.TaskToken) > 0) // has valid task token
+	s.True(len(wft1.Messages) > 0)  // has valid message
+	s.Equal(int64(10), wft1.StartedEventId)
+	s.Equal(int64(9), wft1.Messages[0].GetEventId())
+
+	// unload the shard, this will make last speculative wft disappear from server
+	_, err = s.adminClient.CloseShard(testCtx, &adminservice.CloseShardRequest{
+		ShardId: int32(shardId),
+	})
+	s.NoError(err)
+
+	// send another update wf request (with SAME updateId), this will trigger a new speculative wft
+	go func() {
+		_, _ = s.sendUpdate(tv, "1")
+	}()
+	time.Sleep(time.Second) // sleep 1s to make sure update reached to server
+
+	// poll the new wft (not speculative anymore)
+	wft2, err2 := s.engine.PollWorkflowTaskQueue(testCtx, &workflowservice.PollWorkflowTaskQueueRequest{
+		Namespace: s.namespace,
+		TaskQueue: tv.TaskQueue(),
+	})
+	s.NoError(err2)
+	s.NotNil(wft2)
+	s.True(len(wft2.TaskToken) > 0) // has valid task token
+	s.True(len(wft2.Messages) > 0)  // has valid message
+	s.Equal(int64(10), wft2.StartedEventId)
+	s.Equal(int64(9), wft2.Messages[0].GetEventId())
+
+	// now try to complete 1st speculative wft, it should succeed
+	commands := s.acceptUpdateCommands(tv, "1")
+	commands = append(commands, &commandpb.Command{
+		CommandType: enumspb.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK,
+		Attributes: &commandpb.Command_ScheduleActivityTaskCommandAttributes{ScheduleActivityTaskCommandAttributes: &commandpb.ScheduleActivityTaskCommandAttributes{
+			ActivityId:             tv.ActivityID("13"),
+			ActivityType:           tv.ActivityType(),
+			TaskQueue:              tv.TaskQueue(),
+			ScheduleToCloseTimeout: tv.InfiniteTimeout(),
+		}},
+	})
+	_, err = s.engine.RespondWorkflowTaskCompleted(testCtx, &workflowservice.RespondWorkflowTaskCompletedRequest{
+		Namespace:             s.namespace,
+		TaskToken:             wft1.TaskToken,
+		Commands:              commands,
+		Messages:              s.acceptUpdateMessages(tv, wft1.Messages[0], "1"),
+		ReturnNewWorkflowTask: true,
+	})
+	s.NoError(err) // Staled speculative WFT should be accepted because it has same scheduled_id / started_id and the accepted message is valid (same update_id)
```

Comment:
```suggestion
	s.NoError(err) // Stale speculative WFT should be accepted because it has same scheduled_id / started_id and the accepted message is valid (same update_id)
```